certexchange/polling/subscriber.go

```diff
@@ -0,0 +1,176 @@
+
+// A polling Subscriber will continuously poll the network for new finality certificates.
+type Subscriber struct {
+	certexchange.Client
+
+	Store               *certstore.Store
+	SignatureVerifier   gpbft.Verifier
+	InitialPollInterval time.Duration
+	MaximumPollInterval time.Duration
+	MinimumPollInterval time.Duration
+
+	peerTracker *peerTracker
+	poller      *Poller
+	discoverCh  <-chan peer.ID
+
+	wg   sync.WaitGroup
+	stop context.CancelFunc
+}
+
+func (s *Subscriber) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.stop = cancel
+
+	var err error
+
+	s.peerTracker = newPeerTracker()
+	s.poller, err = NewPoller(ctx, &s.Client, s.Store, s.SignatureVerifier)
+	if err != nil {
+		return err
+	}
+
+	s.discoverCh, err = discoverPeers(ctx, s.Host, s.NetworkName)
+	if err != nil {
+		return err
+	}
+
+	s.wg.Add(1)
+	go func() {
+		defer func() {
+			// in case we return early, cancel.
+			s.stop()
+			// and wait for discovery to exit.
+			for range s.discoverCh {
+			}
+
+			// then we're done
+			s.wg.Done()
+		}()
+
+		if err := s.run(ctx); err != nil && ctx.Err() == nil {
+			s.Log.Errorf("polling certificate exchange subscriber exited early: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+func (s *Subscriber) Stop() error {
+	if s.stop != nil {
+		s.stop()
+		s.wg.Wait()
+	}
+
+	return nil
+}
+
+func (s *Subscriber) run(ctx context.Context) error {
+	timer := clk.Timer(s.InitialPollInterval)
+	defer timer.Stop()
+
+	predictor := newPredictor(
+		s.MinimumPollInterval,
+		s.InitialPollInterval,
+		s.MaximumPollInterval,
+	)
+
+	for ctx.Err() == nil {
+		select {
+		case p := <-s.discoverCh:
+			s.peerTracker.peerSeen(p)
+		case pollTime := <-timer.C:
+			// First, see if we made progress locally. If we have, update
+			// interval prediction based on that local progress. If our interval
+			// was accurate, we'll keep predicting the same interval and we'll
+			// never make any network requests. If we stop making local
+			// progress, we'll start making network requests again.
+			progress, err := s.poller.CatchUp(ctx)
+			if err != nil {
+				return err
+			}
+			// Otherwise, poll the network.
+			if progress == 0 {
+				progress, err = s.poll(ctx)
+				if err != nil {
+					return err
+				}
+			}
+
+			nextInterval := predictor.update(progress)
+			nextPollTime := pollTime.Add(nextInterval)
+			delay := max(clk.Until(nextPollTime), 0)
+			s.Log.Infof("predicted interval is %s (waiting %s)", nextInterval, delay)
+			timer.Reset(delay)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return ctx.Err()
+}
+
+func (s *Subscriber) poll(ctx context.Context) (uint64, error) {
+	var (
+		misses []peer.ID
+		hits   []peer.ID
+	)
+
+	peers := s.peerTracker.suggestPeers()
+	start := s.poller.NextInstance
+	s.Log.Debugf("polling %d peers for instance %d", len(peers), s.poller.NextInstance)
+	for _, peer := range peers {
+		oldInstance := s.poller.NextInstance
+		res, err := s.poller.Poll(ctx, peer)
```

Comment:
Note that we need a timeout here to prevent peers from stalling us. It can be an issue for the future.

Reply:
The "client" can (and should) be configured with a timeout. Although varying the timeout over time given the average request time would be a nice touch.